core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -1558,51 +1576,165 @@ public void testIndexWriterInfoStream() {
         }
     }
 
-    public void testSeqNoAndLocalCheckpoint() {
-        int opCount = randomIntBetween(1, 10);
-        long seqNoCount = -1;
-        for (int op = 0; op < opCount; op++) {
-            final String id = randomFrom("1", "2", "3");
-            ParsedDocument doc = testParsedDocument(id, id, "test", null, -1, -1, testDocumentWithTextField(), B_1, null);
-            if (randomBoolean()) {
-                final Engine.Index index = new Engine.Index(newUid(id), doc,
-                        SequenceNumbersService.UNASSIGNED_SEQ_NO,
-                        rarely() ? 100 : Versions.MATCH_ANY, VersionType.INTERNAL,
-                        PRIMARY, System.currentTimeMillis());
-
-                try {
-                    engine.index(index);
-                } catch (VersionConflictEngineException e) {
-                    // OK
-                }
-                if (index.seqNo() != SequenceNumbersService.UNASSIGNED_SEQ_NO) {
-                    seqNoCount++;
-                    Engine.Index replica = new Engine.Index(index.uid(), index.parsedDoc(), index.seqNo(),
-                            index.version(), VersionType.EXTERNAL, REPLICA, System.currentTimeMillis());
-                    replicaEngine.index(replica);
+    public void testSeqNoAndCheckpoints() throws IOException {
```

Comment:
I'm not sure about this test - I know that it's an extension of the previous version, so my doubts relate to that as well. The test is very complex and it also tries to simulate the interaction between primary and replica but in a limited way (no global checkpoint updates).  

I think in this level of the code we need to make sure that the engine behaves as expected with no concern to cross engine logic. We should just start an engine , index into it and see that the checkpoints are updated as expected and are persisted and recovered. We can randomly choose a "mode" between replica and primary in the beginning of the method, or if it just ends up being too much of a mess, just have two tests. WDYT?